### DISCUSSION TITLE
Suggestion to loosen type specification on Blueprint::HTML::Utils#plain(content : String) from String to any

### DISCUSSION THREAD

**Discussion by robacarp** [stephannv/blueprint#76] [Ideas]
I'm normally in favor of a strictly typed interface, especially at the boundary layer of a shard. But I wonder if the expressiveness here might actually benefit from loosening it up a bit where content is passed into a tag renderer.

For example [this method declaration](https://github.com/stephannv/blueprint/blob/main/src/blueprint/html/utils.cr#L2-L4) has a strict argument type signature of `String` where it's not actually needed:

```crystal
  private def plain(content : String) : Nil
    append_to_buffer(content)
  end
```

That results in needing to explicitly `.to_s` a number, where that's unlikely to be actually necessary:

```crystal
   # explicit cast of data to string, because the compiler will complain without it
   plain query.count.to_s
```

What do you think about adding an overload method which calls .to_s on the `content`?

```crystal
  private def plain(content) : Nil
    append_to_buffer(content.to_s)
  end
```

The `.to_s` is naïve, but it'll result in not needing to manually call `.to_s` on every non-string which has a reasonable .to_s implementation. Consider this:

```
    h2 do
      plain "Showing "
      plain query.current_range.begin.to_s
      plain " to "
      plain query.current_range.end.to_s
      plain " of "
      plain query.total_hits.to_s
      plain " results"
    end
```

This type of desired result is the worst case for this style of html templating engine like blueprint or luckytemplate -- it requires a lot of overhead to compile a simple string, and in this case each data point being injected requires an explicit `.to_s` as well.

**Comment by stephannv** [comment]
@robacarp it makes sense. I will release a new version today with this change. Thanks for the suggestion.

**Reply by stephannv** [reply]
Done @robacarp. You can try it installing 0.10.0 version: https://github.com/stephannv/blueprint/releases/tag/v0.10.0. 

Let me know if this worked for you. 

Another thing that could work, is using String.interpolate, I didn't try but it should work fine:
```crystal
plain String.interpolation(
  "Showing ", 
  query.current_range.begin,
  " to ",
  query.current_range.end,
  " of ",
  query.total_hits,
  " results"
)
```

**Reply by robacarp** [reply]
Interesting, I didn't know about String#interpolation. For that particular stanza, I ended up with this: 

```crystal
    h2 do
      plain "Showing " +
        query.current_range.begin.to_s +
        " to " +
        query.current_range.end.to_s +
        " of " +
        query.total_hits.to_s +
        " results"
    end
```

Ultimately I think there's no elegant way to do this assembly within the template, and I'll probably reach for a presenter pattern instead.

**Reply by robacarp** [reply]
Thanks for the quick turnaround on the update, it also allows this syntax:

```crystal
    h2 do
      plain "Showing "
      plain query.current_range.begin
      plain " to "
      plain query.current_range.end
      plain " of "
      plain query.total_hits
      plain " results"
    end
```

I think I like it the best, because it doesn't feel like it needs the hanging indent.

**Reply by stephannv** [reply]
You're welcome. If you have any suggestions for improving Blueprint developer experience, or new feature, let me know.
